### PR TITLE
[self-profiler] add selfprofiling to llvm

### PR DESCRIPTION
**Pull request by andjo403**
using pass name as event id and add additional data with name of module, function …

image

r? @michaelwoerister or @wesleywiser

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Thanks for the PR, @andjo403! I'll try to take a closer look at this some time later this week.

File: src/librustc_data_structures/profiling.rs
Comment (wesleywiser) [inline src/librustc_data_structures/profiling.rs]:
```suggestion
    llvm_self_profiler: *mut c_void,
```

File: src/librustc_data_structures/profiling.rs
Comment (wesleywiser) [inline src/librustc_data_structures/profiling.rs]:
```suggestion
    let llvm_self_profiler = &mut *(timing_stack as *mut LlvmSelfProfiler<'_>);
```

File: src/librustc_data_structures/profiling.rs
Comment (wesleywiser) [inline src/librustc_data_structures/profiling.rs]: Same thing here

File: src/librustc_data_structures/profiling.rs
Comment (wesleywiser) [inline src/librustc_data_structures/profiling.rs]: We shouldn't copy the value of the separator byte into this file. It should be public so we can just do `measureme::event_id::SEPARATOR_BYTE` here.

File: src/librustc_data_structures/profiling.rs
Comment (wesleywiser) [inline src/librustc_data_structures/profiling.rs]: When you're ready to merge, don't forget to update the `-Z help` docs 🙂

**Review by wesleywiser — CHANGES_REQUESTED** [review]
I have some minor feedback about the self-profiler changes. I don't know enough about LLVM to review those parts so I'll leave that to @michaelwoerister.

File: src/librustc_data_structures/profiling.rs
Comment (andjo403) [reply src/librustc_data_structures/profiling.rs]: do not know why I was sure that I needed to do an update in measureme to get it public :)

File: src/librustc_data_structures/profiling.rs
Comment (andjo403) [reply src/librustc_data_structures/profiling.rs]: waited with that until the PR that I build on to is merged so I can rebase and get your changes :) but leaving this as unresolved so I do not forget.

File: src/librustc_data_structures/profiling.rs
Comment (michaelwoerister) [inline src/librustc_data_structures/profiling.rs]: Is there ever more than one argument? If not you can use the `EventIdBuilder::from_label_and_arg`.

File: src/librustc_data_structures/profiling.rs
Comment (michaelwoerister) [inline src/librustc_data_structures/profiling.rs]: I think `LlvmSelfProfiler` and all of the related code should be moved to `librustc_codegen_llvm`, e.g. into a file `librustc_codegen_llvm/back/profiling.rs`.

File: src/librustc_data_structures/profiling.rs
Comment (michaelwoerister) [inline src/librustc_data_structures/profiling.rs]: I have a PR in the works that allows for caching non-`'static` strings in the `SelfProfiler` so this separate cache won't be needed soon.

**Review by michaelwoerister — COMMENTED** [review]
I did a first review pass. Looks great! I want to take a closer look before merging. For now, could you move the LLVM specific code to `librustc_codegen_llvm` (as mentioned below).

File: src/librustc_data_structures/profiling.rs
Comment (andjo403) [reply src/librustc_data_structures/profiling.rs]: for LazyCallGraph::SCC there can be up to 10 args (if more they are truncated by llvm) do not know if it is interesting to see all the functions in the SCC (and in that case make a new getName function in llvm that returns all not only 10) or if it is enough to print the first function.  see around https://github.com/rust-lang/llvm-project/blob/a6f4c1bb07d58df5956d2c49be68547143f77b18/llvm/include/llvm/Analysis/LazyCallGraph.h#L447 for more info.

File: src/librustc_data_structures/profiling.rs
Comment (andjo403) [reply src/librustc_data_structures/profiling.rs]: will look in to doing this next week

File: src/librustc_data_structures/profiling.rs
Comment (andjo403) [reply src/librustc_data_structures/profiling.rs]: that sounds great. this was kind of a hack to avoiding the file to get larger than 1GB

File: src/librustc_data_structures/profiling.rs
Comment (michaelwoerister) [reply src/librustc_data_structures/profiling.rs]: No rush

File: src/librustc_data_structures/profiling.rs
Comment (michaelwoerister) [reply src/librustc_data_structures/profiling.rs]: Yeah we'll want to add a `EventIdBuilder::from_label_and_args(label: _, args: &[_])` method at some point anyway. For now encoding things directly is fine.

File: src/librustc_data_structures/profiling.rs
Comment (andjo403) [reply src/librustc_data_structures/profiling.rs]: have started looking at this and have a question.
now I have a get_llvm_selfprofiler() function in SelfProfilerRef but that do not work after the move. but some how I need to get a selfprofiler to save in LlvmSelfProfiler and I need to check if the llvm event is wanted. shall I make profiler public(or a get function) and add a llvm_recording_enabled in SelfProfilerRef? or shall I have a get_profiler_when_llvm_recording_enabled that returns a option? I feel like neither is a good solution.

**Comment by andjo403** [comment]
have fixed the comments now

**Comment by michaelwoerister** [comment]
Looks good to me, thanks @andjo403! Now just have to wait for #67954 to land.

**Comment by andjo403** [comment]
shall there be some check or documentation that llvm event recording only works when `-Znew-llvm-pass-manager=y` is also added can maybe be confusing when there is no llvm events in the logs?

**Comment by wesleywiser** [comment]
That's a good idea and makes sense to me. I'm not sure we should error out in that case but a warning or similar seems appropriate to me.

**Comment by michaelwoerister** [comment]
Yes, a warning makes sense to me too.

**Comment by andjo403** [comment]
how do I add a warning? was not that easy to find some other code that use it.
`rg warning` did give to many results :)

**Comment by bjorn3** [comment]
You could use [`sess.warn()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_session/struct.Session.html#method.warn).

File: src/librustc_data_structures/profiling.rs
Comment (andjo403) [reply src/librustc_data_structures/profiling.rs]: have remove the string_cache and uses the new get_or_alloc_cached_string one down side with the change is that I need to demangle all the time but before I only did it when the string was cached.

**Comment by andjo403** [comment]
added a warning if the new passmanager is not used and llvm events is requested.
do not know if it is a good place where I trigger the warning.

**Comment by wesleywiser** [comment]
Placement of the warning seems fine to me

**Comment by andjo403** [comment]
the PR that this depended on have merged to master and I have now rebase onto master.
so this PR is now ready for final review.

**Review by wesleywiser — APPROVED** [review]

**Pull request merged**
